### PR TITLE
Classification Metric UDFs

### PR DESCRIPTION
**Pull request by atavory**
Following #12732 and talks with @rongrong, I'd like to request pulling a Precision/Recall UDF.

```
.. function:: classification_miss_rate(buckets, y, x, weight) -> array<double>

     Computes the miss-rate part of the receiver operator curve with up to ``buckets`` number of buckets. Returns
    an array of miss-rate values. ``y`` should be a boolean outcome value; ``x`` should be predictions, each
    between 0 and 1; ``weight`` should be non-negative values, indicating the weight of the instance.

     To get an ROC map, use this in conjunction with :func:`classification_recall`:

     .. code-block:: none

         MAP(classification_recall(200, outcome, prediction), classification_miss_rate(200, outcome, prediction))

 .. function:: classification_precision(buckets, y, x) -> array<double>

     This function is equivalent to the variant of
    :func:`classification_precision` that takes a ``weight``, with a per-item weight of ``1``.

 .. function:: classification_precision(buckets, y, x, weight) -> array<double>

     Computes the precision part of the precision-recall curve with up to ``buckets`` number of buckets. Returns
    an array of precision values. ``y`` should be a boolean outcome value; ``x`` should be predictions, each
    between 0 and 1; ``weight`` should be non-negative values, indicating the weight of the instance.

     To get a map of recall to precision, use this in conjunction with :func:`classification_recall`:

     .. code-block:: none

         MAP(classification_recall(200, outcome, prediction), classification_precision(200, outcome, prediction))

 .. function:: classification_precision(buckets, y, x) -> array<double>

     This function is equivalent to the variant of
    :func:`classification_precision` that takes a ``weight``, with a per-item weight of ``1``.

 .. function:: classification_recall(buckets, y, x, weight) -> array<double>

     Computes the recall part of the precision-recall curve or the receiver operator charateristic curve
    with up to ``buckets`` number of buckets. Returns an array of recall values.
    ``y`` should be a boolean outcome value; ``x`` should be predictions, each
    between 0 and 1; ``weight`` should be non-negative values, indicating the weight of the instance.

     To get a map of recall to precision, use this in conjunction with :func:`classification_recall`:

     .. code-block:: none

         MAP(classification_recall(200, outcome, prediction), classification_precision(200, outcome, prediction))

 .. function:: classification_recall(buckets, y, x) -> array<double>

     This function is equivalent to the variant of
    :func:`classification_recall` that takes a ``weight``, with a per-item weight of ``1``.

 .. function:: classification_thresholds(buckets, y, x) -> array<double>

     Computes the thresholds part of the precision-recall curve with up to ``buckets`` number of buckets. Returns
    an array of thresholds. ``y`` should be a boolean outcome value; ``x`` should be predictions, each
    between 0 and 1.

     To get a map of thresholds to precision, use this in conjunction with :func:`classification_precision`:

     .. code-block:: none

         MAP(classification_thresholds(200, outcome, prediction), classification_precision(200, outcome, prediction))

     To get a map of thresholds to recall, use this in conjunction with :func:`classification_recall`:

     .. code-block:: none

         MAP(classification_thresholds(200, outcome, prediction), classification_recall(200, outcome, prediction))
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateSerializer.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        int requiredBytes = SizeOf.SIZE_OF_INT + // null
+                SizeOf.SIZE_OF_INT; // has histograms;
+        if (state.getTrueWeights() != null) {
+            requiredBytes += state.getTrueWeights().getRequiredBytesForSerialization();
+            requiredBytes += state.getFalseWeights().getRequiredBytesForSerialization();
+        }
+
+        SliceOutput sliceOut = Slices.allocate(requiredBytes).getOutput();
+        sliceOut.appendInt(state.getNull() ? 1 : 0);
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateSerializer.java]: Nit: You could `appendByte` here

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    public long estimatedInMemorySize()
+    {
+        return INSTANCE_SIZE + SizeOf.sizeOf(weights);
+    }
+
+    public void set(double value, double weight)
+    {
+        checkArgument(value >= min, "value must not be smaller than min");
+        checkArgument(value <= max, "value must not be larger than max");
+        final int ind = Math.min(
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]: I would have a little helper function for here and add and call it something like `getIndex`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    public long estimatedInMemorySize()
+    {
+        return INSTANCE_SIZE + SizeOf.sizeOf(weights);
+    }
+
+    public void set(double value, double weight)
+    {
+        checkArgument(value >= min, "value must not be smaller than min");
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]:
```suggestion
        checkArgument(value >= min && value <= max, "value must be within range [%s, %s]", min, max);
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    public void add(double value)
+    {
+        add(value, 1);
+    }
+
+    public void add(double value, double weight)
+    {
+        checkArgument(value >= min, "value must not be smaller than min");
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]:
```suggestion
        checkArgument(value >= min && value <= max, "value must be within range [%s, %s]", min, max);
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        checkArgument(value <= max, "value must not be larger than max");
+        final int ind = Math.min(
+                (int) (numBuckets * (value - min) / (max - min)),
+                numBuckets - 1);
+        weights[ind] += weight;
+    }
+
+    public void mergeWith(FixedNumericHistogram other)
+    {
+        checkArgument(numBuckets == other.numBuckets, "numBuckets must be equal");
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]: I would indicate numbers that mismatch in the message

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FixedNumericHistogram
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]: I would call this `FixedDoubleHistogram`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        checkArgument(numBuckets == other.numBuckets, "numBuckets must be equal");
+        checkArgument(min == other.min, "min must be equal");
+        checkArgument(max == other.max, "max must be equal");
+
+        for (int i = 0; i < numBuckets; ++i) {
+            weights[i] += other.weights[i];
+        }
+    }
+
+    public SortedMap<Double, Double> getBuckets()
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]: I would consider having an object that represents a bin, and returning an `Iterable<Bin>` here.

We don't need to change this implementation to internally use Bin objects.  The flat memory we are using is already great.  We can instead fake it by mutating a singleton `Bin` and mutate it with the next bin's numbers when `iterator#next` is called.  This way we avoid all of the autoboxing and the need to create a map, since all we want to do is iterate over the bins in ascending order.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    {
+        SortedMap<Double, Double> result = new TreeMap<>();
+        for (int i = 0; i < numBuckets; i++) {
+            result.put(min + (double) (i) / numBuckets * (max - min), weights[i]);
+        }
+        return result;
+    }
+
+    public FixedNumericHistogram copy()
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]: I would instead have a copy constructor.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+        }
+
+        final Double totalFalseWeight = state.getFalseWeights().getBuckets().values().stream()
+                .mapToDouble(Double::doubleValue)
+                .sum();
+
+        SortedMap<Double, Double> result = new TreeMap<>();
+        Double runningWeight = 0.0;
+        Double runningTrueWeight = 0.0;
+        Iterator<Map.Entry<Double, Double>> trueIt = state.getTrueWeights().getBuckets().entrySet().iterator();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: The suggestion on using an iterator rather than a map may also make this part a bit more clear, since the key and values have specific meanings which you can't know just by looking at the generics.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    private PrecisionRecallAggregation() {}
+
+    private static Double safeDiv(Double nom, Double denom)
+    {
+        return denom == 0.0 ? 1.0 : nom / denom;
+    }
+
+    private static SortedMap<Double, Double> getResultMap(@AggregationState PrecisionRecallState state)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: So if we go with the iterator approach I mentioned above, we can change this method to not return a map and directly write out the curve, to avoid repackaging in a map which we'll then just iterate over to serialize.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            final Double precision = entry.getValue();
+            DoubleType.DOUBLE.writeDouble(entryBuilder, recall);
+            DoubleType.DOUBLE.writeDouble(entryBuilder, precision);
+        }
+        out.closeEntry();
+    }
+
+    private PrecisionRecallAggregation() {}
+
+    private static Double safeDiv(Double nom, Double denom)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]:
```suggestion
    private static Double safeDivide(Double nom, Double denom)
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+    @InputFunction
+    public static void input(
+            @AggregationState PrecisionRecallState state,
+            @SqlType(StandardTypes.BIGINT) long numBins,
+            @SqlType(StandardTypes.BOOLEAN) boolean outcome,
+            @SqlType(StandardTypes.DOUBLE) double pred,
+            @SqlType(StandardTypes.DOUBLE) double weight)
+    {
+        if (state.getTrueWeights() == null) {
+            state.setTrueWeights(new FixedNumericHistogram((int) (numBins), 0.0, 1.0));
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: nit: let's add constants `MIN_BUCKET_VALUE`, `MAX_BUCKET_VALUE`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateFactory.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    {
+        return GroupedState.class;
+    }
+
+    public static class GroupedState
+            extends AbstractGroupedAccumulatorState
+            implements PrecisionRecallState
+    {
+        private boolean nullSet;
+        private final ObjectBigArray<FixedNumericHistogram> trueWeights = new ObjectBigArray<>();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateFactory.java]: This is optional, and somewhat lengthy.  I leave it up to you if you want to do this.

Presto, as you know, holds everything in memory, and won't spill for aggregations.  It also runs within the JVM and so it's sensitive to GC latency.  In the past, some aggregation functions which use `ObjectBigArray` of individual states (per group) have caused issues (the groupings can be in the billions).  Some aggregation functions have been rewritten in such a way that their state for grouped aggregation is totally flat--meaning, there's a constant order of objects created to represent the grouped state, not order of N.  This is done to make them more GC-friendly.  Typically, this involves stuffing the group id into the state itself somehow (for example, `array_agg` does this by interleaving the individual array aggregations via a linked list, which we then traverse over to reconstruct the per-grouping array aggregation).

It should be relatively easy to do something similar for this, because we know exactly how big each histogram is.  So we could have some sort of GroupedHistogram, which takes in the group id plus the value, and we use the group id to find the starting index into a region of an underlying giant (and GC-friendly) array that holds the histogram state for all groups.

It may be a bit of a premature optimization, because I'm not sure how in practice this function may be used, and how large the histograms could grow.  So this is optional feedback.

CC: @wenleix

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 14]
```diff
@@ -77,13 +77,8 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
         return Result.ofPlanNode(getSyntacticOrderJoin(joinNode, context, joinDistributionType));
     }
 
-    public static boolean canReplicate(JoinNode joinNode, Context context)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: What is the purpose of this change?

File: presto-docs/src/main/sphinx/functions/aggregate.rst [hunk 15]
```diff
@@ -317,6 +317,13 @@ Statistical Aggregate Functions
 
         kurtosis(x) = n(n+1)/((n-1)(n-2)(n-3))sum[(x_i-mean)^4]/stddev(x)^4-3(n-1)^2/((n-2)(n-3))
 
+.. function:: precision_recall(buckets, outcome, pred, weight) -> map<double, double>
+
+    Computes the precision-recall curve with up to ``buckets`` number of buckets. Returns a map mapping
+    recall values to precision values.
+
+    :func:`precision_recall` that takes a ``weight``, with a per-item weight of ``1``.
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/functions/aggregate.rst]: Looks like this is broken or incomplete

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.TreeMap;
+
+@AggregationFunction("precision_recall")
+@Description("Computes precision-recall curves")
+public final class PrecisionRecallAggregation
+{
+    @InputFunction
+    public static void input(
+            @AggregationState PrecisionRecallState state,
+            @SqlType(StandardTypes.BIGINT) long numBins,
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: I would make it `INTEGER` and use an `int`.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+import java.util.TreeMap;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FixedNumericHistogram
+{
+    private static final byte FORMAT_TAG = 0;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FixedNumericHistogram.class).instanceSize();
+
+    private final int numBuckets;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedNumericHistogram.java]: I would call it `bucketCount`, we try to avoid abbreviations in the code

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java (hunk 16 above)
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: I would be nice to add a getter to retrieve the number of bins in a histogram, and then validate that it's the same as what's passed in here (since it's supposed to be constant)

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+            @SqlType(StandardTypes.BOOLEAN) boolean outcome,
+            @SqlType(StandardTypes.DOUBLE) double pred,
+            @SqlType(StandardTypes.DOUBLE) double weight)
+    {
+        if (state.getTrueWeights() == null) {
+            state.setTrueWeights(new FixedNumericHistogram((int) (numBins), 0.0, 1.0));
+            state.setFalseWeights(new FixedNumericHistogram((int) (numBins), 0.0, 1.0));
+        }
+
+        if (pred < 0 || pred > 1 || weight < 0.0) {
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: Couldn't this be performed outside of the function itself?  Perhaps someone would prefer it to throw rather than treat it as null?  But if the default is to throw, one could always perform this transformation outside of the function, however the converse isn't true.

File: presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.type.DoubleType.DOUBLE;
+
+public class TestPrecisionRecallAggregation
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java]: It would be nice to have some tests for invalid inputs.

**Comment by tdcmeehan** [comment]
Does it make sense to separate the aggregation of precision from recall?  You can always `MAP(ARRAY[precision],ARRAY[recall])`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java (hunk 10 above)
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: I actually like using your idea of an iterator here as well, just a composite (zip) iterator.

**Comment by atavory** [comment]
@tdcmeehan, many thanks for your excellent suggestions and comments! I've addressed some, and will continue tomorrow.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java (hunk 18 above)
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: Not sure I get it. It is a violation of the contract of the function, in some sense, just like changing the number of bins between instances (which should be constant). IMHO, these should all be treated the same way. Are you suggesting to use `checkArgument`, or just not check (and whatever will happen is on the user), or return `NULL`?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateFactory.java (hunk 13 above)
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/PrecisionRecallStateFactory.java]: Very interesting point. For precision/recall, I think it is less relevant (esp. before evidence to the contrary). Meaningful precision/recall is typically done by training over large numbers of instances, so the case where there is a large number of partitions, with meaningful precision/recall for each, seems to me less common. Perhaps we could do this PR without it, and then consider?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java (hunk 18 above)
Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: We can throw a PrestoException which indicates a data quality issue.  i.e. we can throw `new PrestionException(INVALID_FUNCTION_ARGUMENT, ...)` if these things occur.  If you don't care about that thought, you can wrap the function in SQL with `TRY`, which will return null if that gets thrown.

**Comment by atavory** [comment]
@tdcmeehan Thanks for your comments!

I especially liked the last one, about using a `MAP` of separate functions. This solves a problem I've seen with other APIs, where there are various different metrics associated with such curves, and some parameters make sense in only some cases, and only when some metrics are even needed (see https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_fscore_support.html, https://scikit-learn.org/stable/auto_examples/model_selection/plot_precision_recall.html, https://en.wikipedia.org/wiki/Matthews_correlation_coefficient, https://en.wikipedia.org/wiki/Youden%27s_J_statistic, https://en.wikipedia.org/wiki/Cohen%27s_kappa). 

By dividing the aggregations into `classification_precision`, `classification_recall`, and `classification_thresholds`, it is possible to use the combination the user wants (e.g., to find precision-recall curves, or precision-threshold curves). It should also be easy to add further metrics if needed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 14 above)
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: Sorry! Mistake.

File: presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java (hunk 19 above)
Comment (atavory) [reply presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java]: Done

**Comment by tdcmeehan** [comment]
Just skimmed, but looks good!  Will give this a full review tomorrow.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ClassificationMissRateAggregation.java [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+        while (resultsIterator.hasNext()) {
+            final BucketResult result = resultsIterator.next();
+            DoubleType.DOUBLE.writeDouble(
+                    entryBuilder,
+                    result.runningFalseWeight / result.totalTrueWeight);
+        }
+        out.closeEntry();
+    }
+
+    private ClassificationMissRateAggregation() {}
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ClassificationMissRateAggregation.java]: By convention, I would put this at the top rather than bottom.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/ClassificationPrecisionAggregation.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+{
+    @OutputFunction("array(double)")
+    public static void output(@AggregationState PrecisionRecallState state, BlockBuilder out)
+    {
+        Iterator<PrecisionRecallAggregation.BucketResult> resultsIterator = getResultsIterator(state);
+
+        BlockBuilder entryBuilder = out.beginBlockEntry();
+        while (resultsIterator.hasNext()) {
+            final BucketResult result = resultsIterator.next();
+            final Double remainingTrueWeight = result.totalTrueWeight - result.runningTrueWeight;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/ClassificationPrecisionAggregation.java]: We use small `d` `double`s for simple calculations.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FixedDoubleHistogram.class).instanceSize();
+
+    private final int bucketCount;
+    private final double min;
+    private final double max;
+    private final double[] weights;
+
+    public class Bin
+    {
+        public final Double left;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: Let's use small D `double` wherever possible

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java [hunk 23]
```diff
@@ -0,0 +1,180 @@
...
+            @Override
+            public boolean hasNext()
+            {
+                return currentIndex < bucketCount;
+            }
+
+            @Override
+            public Bin next()
+            {
+                return new Bin(
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: We should throw `NoSuchElementException` if we attempt `next()` but there's no more elements

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 24]
```diff
@@ -0,0 +1,192 @@
...
+
+    protected PrecisionRecallAggregation() {}
+
+    protected static Iterator<BucketResult> getResultsIterator(@AggregationState PrecisionRecallState state)
+    {
+        if (state.getTrueWeights() == null) {
+            return Collections.<BucketResult>emptyList().iterator();
+        }
+
+        final Double totalTrueWeight = Streams.stream(state.getTrueWeights().iterator())
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: This can be small D `double`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+
+            @Override
+            public void remove()
+            {
+                throw new UnsupportedOperationException();
+            }
+        };
+    }
+
+    private static final double DEFAULT_WEIGHT = 1.0;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: By convention, these all go to the very top of the class

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 26]
```diff
@@ -0,0 +1,192 @@
...
+            @Override
+            public boolean hasNext()
+            {
+                return trueIt.hasNext() && totalTrueWeight > runningTrueWeight;
+            }
+
+            @Override
+            public BucketResult next()
+            {
+                final FixedDoubleHistogram.Bin trueResult = trueIt.next();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: We should throw `NoSuchElementException` if we attempt `next()` but there's no more elements.  It's minor, and although `trueIt` may throw this, it should be thrown here rather than in the inner iterator.

File: presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+                functionManager.lookupFunction(
+                        this.functionName,
+                        fromTypes(BIGINT, BOOLEAN, DOUBLE, DOUBLE)));
+    }
+
+    @Test
+    public void testNegativeWeight()
+    {
+        String error = "";
+        try {
```
Comment (tdcmeehan) [inline presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java]: I think we can omit the check on the specific text of the exception and wrap the `assertAggregation` with an `assertThrows(PrestoException.class, () -> assertAggregation(...))`

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+{
+    private static final byte FORMAT_TAG = 0;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FixedDoubleHistogram.class).instanceSize();
+
+    private final int bucketCount;
+    private final double min;
+    private final double max;
+    private final double[] weights;
+
+    public class Bin
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: This should be public static.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 29]
```diff
@@ -0,0 +1,192 @@
...
+        }
+        if (state.getTrueWeights() != null && otherState.getTrueWeights() != null) {
+            state.getTrueWeights().mergeWith(otherState.getTrueWeights());
+            state.getFalseWeights().mergeWith(otherState.getFalseWeights());
+        }
+    }
+
+    protected static class BucketResult
+    {
+        public final Double totalTrueWeight;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: Just a reminder on the use of `Double`.  The reason I'm saying this is two fold:

1) It incurs object allocation overhead that didn't otherwise exist 
2) It makes one read this code wonder if there's a null case, which isn't possible for primitive double

From what I've read, I don't see any case where any of these could be null.

File: presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java (hunk 27 above)
Comment (atavory) [reply presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java]: Not sure I follow. If this case happens, I'd like an exception thrown which contains this text, as it is indicative of the problem that actually happened. This is what the test does.

Comment (tdcmeehan) [reply presto-main/src/test/java/com/facebook/presto/operator/TestPrecisionRecallAggregation.java]: IMO it's sufficient to just test that it throws.  If we really want to check on the exception text, then I'd do it differently:

```java
        try {
            assertAggregation(
                    precisionRecallFunction,
                    0.0,
                    createLongsBlock(Long.valueOf(200)),
                    createBooleansBlock(Boolean.valueOf(true)),
                    createDoublesBlock(Double.valueOf(0.2)),
                    createDoublesBlock(Double.valueOf(-0.2)));
            fail("Expected PrestoException");
        }
        catch (PrestoException e) {
            assertTrue(e.getText().toLowerCase(Locale.ENGLISH).contains("weight"));
            assertTrue(e.getText().toLowerCase(Locale.ENGLISH).contains("negative"));
        }
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java [hunk 30]
```diff
@@ -0,0 +1,184 @@
...
+
+            @Override
+            public void remove()
+            {
+                throw new UnsupportedOperationException();
+            }
+        };
+    }
+
+    private int getIndexForValue(Double value)
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: I'm fine with the tests having big-D doubles but let's remove all of them from main, we don't need the autoboxing.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+                .sum();
+        final double totalFalseWeight = Streams.stream(state.getFalseWeights().iterator())
+                .mapToDouble(c -> c.weight)
+                .sum();
+
+        return new Iterator<BucketResult>()
+        {
+            Iterator<FixedDoubleHistogram.Bin> trueIt = state.getTrueWeights().iterator();
+            Iterator<FixedDoubleHistogram.Bin> falseIt = state.getFalseWeights().iterator();
+            Double runningFalseWeight = 0.0;
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]: Here as well

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java (hunk 30 above)
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: Thanks for pointing this out - am removing them from the tests as well.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+            public boolean hasNext()
+            {
+                return currentIndex < bucketCount;
+            }
+
+            @Override
+            public Bin next()
+            {
+                if (currentIndex == bucketCount) {
+                    throw new UnsupportedOperationException();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]:
```suggestion
                    throw new NoSuchElementException();
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java [hunk 33]
```diff
@@ -0,0 +1,195 @@
...
+            public boolean hasNext()
+            {
+                return trueIt.hasNext() && totalTrueWeight > runningTrueWeight;
+            }
+
+            @Override
+            public BucketResult next()
+            {
+                if (!trueIt.hasNext() || !falseIt.hasNext()) {
+                    throw new UnsupportedOperationException();
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/PrecisionRecallAggregation.java]:
```suggestion
                    throw new NoSuchElementException();
```

**Comment by tdcmeehan** [comment]
Tagging @rongrong to also take a quick look.  LGTM.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java [hunk 34]
```diff
@@ -0,0 +1,195 @@
...
+    {
+        return min;
+    }
+
+    double getMax()
+    {
+        return max;
+    }
+
+    FixedDoubleHistogram(FixedDoubleHistogram other)
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: I feel that we normally don't do this in Presto but explicitly create a new object with old parameters.

Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java]: @rongrong I've made the class implement `Cloneable`, and implemented `clone`. This is now used instead.
